### ISSUE TITLE
The game freezes on -Dnext

### ISSUE THREAD

**Issue by starry-abyss** [HaxeFlixel/flixel#2008]
- **Flixel version:** dev
- **OpenFL version:** 3.6.1
- **Lime version:** 2.9.1
- **Haxe version:** 3.2.1
- **Affected targets:** neko, windows

________________________________

-Dnext. Only debug builds.
This issue has been around for a while now. I can reproduce it in FlxBunnyMark demo (see below).

________________________________

**Observed behavior:**

**Scenario 1:**
- run the game
- minimize the game window
- restore the window back
- the game freezes, Update and Draw in the flx debugger show: 0

Video capture: http://i.giphy.com/3oriOchXffaK7OSD1m.gif

**Scenario 2:**
- run the game
- open and close the flx debugger with tilda key
- click anywhere inside the game window
- the game freezes (in the video example Focus Lost screen is shown, but defining FLX_NO_FOCUS_LOST_SCREEN doesn't fix the freeze itself)
- open and close the flx debugger with tilda key
- the game unfreezes until the next mouse click and so on

Video capture: http://i.giphy.com/l2Jht0pruoqifEX8Q.gif

**Expected behavior:**

The game should not freeze

**Comment by Cor-cor** [comment]
Just faced same problem on html5 target. Bisected it to this commit: https://github.com/HaxeFlixel/flixel/commit/e802c61bc3151355dbd81ecfbc7b277ea6c07bfb
The game gets paused on focus lost but doesn't automatically unpause afterwards.

**Comment by nazywam** [comment]
Same issue here, html5 export freezes when lost focus, happens only if debug flag is set.
Curiously enough, the `FLX_NO_FOCUS_LOST_SCREEN` flag is set

**Comment by Beeblerox** [comment]
@starry-abyss could you check my fix, please?

**Comment by nazywam** [comment]
@Beeblerox The game is still paused. I.e. you need to resume it from the debugger window, other than that, it seems fixed :)

**Comment by Beeblerox** [comment]
@nazywam do you mean on html5 target?
i haven't checked it yet, because @starry-abyss opened this issue for neko and windows targets

**Comment by nazywam** [comment]
Yup, now that I think about it, it may be a different issue. I'll do some tests and maybe open a new issue for it?

**Comment by Beeblerox** [comment]
Yes. I believe  that this is another issue

**Comment by starry-abyss** [comment]
I'm participating in LD right now, but hang on, I'll check it soon. Thank you for investigating!

**Comment by starry-abyss** [comment]
@Beeblerox It fixes the scenario 2, but not the scenario 1.

**Comment by Beeblerox** [comment]
@starry-abyss are you sure that scenario 1 isn't fixed? can't reproduce on my machine, unfortunately.
could you check one more time please?

**Comment by Beeblerox** [comment]
but now i get new issue:
open debugger, then click on one of the debugger windows (now the window header), then click on visible game area (not covered by debugger windows) -> game freezes (and unfreezes only after i close debugger)

**Comment by starry-abyss** [comment]
@Beeblerox Yes, win7 debug -Dnext neko and windows. Scenario 1 isn't fixed, also I found a variation of it. You can also switch to another window and back to get the freeze.

**Comment by Beeblerox** [comment]
@starry-abyss could you reproduce this issue with this compiled demo?
[bin.zip](https://github.com/HaxeFlixel/flixel/files/652436/bin.zip)
I've compiled it with the latest dev version of HaxeFlixel, lime 2.9.1, openfl 3.6.1 and Haxe 3.4rc

**Comment by Beeblerox** [comment]
sorry, wrong archive (compiled without -Dnext flag)
Here is correct version: 
[bin.zip](https://github.com/HaxeFlixel/flixel/files/652441/bin.zip)

**Comment by Beeblerox** [comment]
hmm, i get different results on windows10 and windows7 with the same demo: my issue reproducable on win7, but not on win10

**Comment by starry-abyss** [comment]
No, your builds work OK! But I use Haxe 3.2.1, and also I don't have 'shaders: off' button in the demo at all

**Comment by Beeblerox** [comment]
button is from latest version of the demo. it shows only when you're using -Dnext flag and native targets

**Comment by starry-abyss** [comment]
OK, I updated the demo, now I have the shaders button, but the game still freezes after gaining the focus again

**Comment by Beeblerox** [comment]
i found one more interesting detail: if i run demo right after compilation (from HaxeDevelop) - then there is no issue i've found, but when i run it manually - then i get this issue. Strange

**Comment by starry-abyss** [comment]
I run from HaxeDevelop. So do you now get the issue I described?

**Comment by Beeblerox** [comment]
no. only when i run it manually

**Comment by starry-abyss** [comment]
Hmm. Doesn't that count as you can now reproduce it yourself?

**Comment by Beeblerox** [comment]
i get only the issue i've described a bit earlier, and i can't reproduce issues you've described

**Comment by Beeblerox** [comment]
i'm sorry. there was some mistake with project configuration and i ran wrong executable manually. Now there are no issues for me at all on neko and windows targets (while running demo from HaxeDevelop or manually)

**Comment by Beeblerox** [comment]
Here is correct (at last) version of compiled demo [bin.zip](https://github.com/HaxeFlixel/flixel/files/652688/bin.zip)

**Comment by starry-abyss** [comment]
You build works OK. I upgraded Haxe to 3.4.0-rc1, and I still get the issue on my PC. Here is my build:
[bin.zip](https://github.com/HaxeFlixel/flixel/files/652718/bin.zip)

**Comment by Beeblerox** [comment]
@starry-abyss i can't even run it on win10, for some reason. i'll try run it on win7 tomorrow

**Comment by starry-abyss** [comment]
I can add that it happens with debug builds only (neko and windows) and not with release ones

**Comment by starry-abyss** [comment]
@Beeblerox I just tested that build on win10 pc, it doesn't crash, and also there is no 2nd scenario freeze, only the 1st!

**Comment by JesseTG** [comment]
I'm being affected by this issue as well, even without `-Dnext`.  Only on HTML5 debug builds, though.  https://github.com/HaxeFlixel/flixel/issues/2008

**Comment by Beeblerox** [comment]
@JesseTG have you tested your code with the latest dev? 
and in what browsers do you have this issue?

**Comment by JesseTG** [comment]
@Beeblerox Chrome and Firefox, haven't tried other browsers.  I *am* using the latest `dev` build of HaxeFlixel.

**Comment by Beeblerox** [comment]
@JesseTG that's strange. because i'm using the same code and couldn't reproduce it in Google Chrome

**Comment by starry-abyss** [comment]
@Beeblerox OK, I have some clue, by commenting out the following line the Scenario 1 is kinda fixed for me (Scenario 2 is not). Objects get updated and drawn, indicators in debugger are non-zero, **BUT** fps after restoring the window becomes ~105 instead of 60 (in the other project with only one object on the screen it's exactly 120).

FlxGame.hx:

```
private function onEnterFrame(_):Void
{
		..........
		
				else if (_state == _requestedState) // don't pause a state switch request
				{
					#if FLX_DEBUG
					debugger.update();
					// If the interactive debug is active, the screen must
					// be rendered because the user might be doing changes
					// to game objects (e.g. moving things around).
					if (debugger.interaction.isActive())
					{
						draw();
					}
					#end
					//return;    <-------------------------- makes the difference for Scenario 1
				}

		..........
}
```

**Comment by MintPaw** [comment]
To try to find a workaround, can anyone point me to the code that pauses when you click in the debug console field so I can try to disable that functionality entirely.

**Comment by starry-abyss** [comment]
@MintPaw No, it's for another issue. See the first post, no workaround for click-case yet, only minimize-restore one.

**Comment by MintPaw** [comment]
@starry-abyss Ah ok, so simply setting FlxG.console.autoPause to false seems to fix things for me.

**Comment by starry-abyss** [comment]
Indeed, the scenario 2 is workarounded with Flx.autoPause = false.

**Comment by starry-abyss** [comment]
Commenting out `stage.addEventListener(FocusEvent.FOCUS_OUT, onFocusLost);` is the clue for Scenario 2. Apparently, we somehow receive this event when clicking in the debugger.

**Comment by starry-abyss** [comment]
Seems like we receive the event from openfl/display/Stage.hx#set_focus().
And it can be send by OpenFl for any InteractiveObject. I think we should analyze, what kind of object lost focus, when we receive the event. At the moment we just assume it can only be received for the whole game, and not for individual objects, which is apparently wrong?

**Comment by tbrosman** [comment]
I'm seeing this issue in specific versions of Flixel with vnext enabled. Here's a table from the Windows builds of my game in debug mode.

Repro steps:

* Click away from the window (the "Play" icon should appear/the game pauses)
* Change focus to the window (the "Play" icon should disappear)

As @starry-abyss described, the debugger can still be accessed via the tilde key.

|Flixel version|Issue|
|---|---|
|4.0.1|Y, keyboard input breaks|
|4.1.1|Y, keyboard input breaks|
|4.2.0|Y, everything freezes|
|4.2.1|Y, everything freezes|

This table also applies to Android (though I haven't verified the keyboard part for those builds).

**Comment by Gama11** [comment]
I'm pretty sure this is no longer an issue with OpenFL 8 (#2136). :) Not really interested in investing the time to find a workaround for OpenFL 3.6.1.
